maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,253 @@
+        """Push business to next step.
+
+        Args:
+            tick (int): Current tick to process.
+        """
+        self._tick = tick
+
+        # Update all live VMs memory utilization.
+        self._update_vm_util()
+        # Update all PM memory utilization.
+        self._update_pm_util()
+        # TODO
+        # Generate VM requirement events from data file.
+        # It might be implemented by a for loop to process VMs in each tick.
+        vm_required_evt = self._event_buffer.gen_cascade_event(tick, Events.REQUIREMENTS, payload=None)
+        self._event_buffer.insert_event(vm_required_evt)
+
+    def get_metrics(self) -> dict:
+        """Get current environment metrics information.
+
+        Returns:
+            dict: Metrics information.
+        """
+
+        return DocableDict(
+            metrics_desc,
+            energy_consumption=self._energy_consumption,
+            success_requirements=self._success_requirements,
+            failed_requirements=self._failed_requirements,
+            total_latency=self._total_latency
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(Events.REQUIREMENTS, self._on_vm_required)
+        self._event_buffer.register_event_handler(Events.FINISHED, self._on_vm_finished)
+
+        # Generate decision event.
+        self._event_buffer.register_event_handler(MaroEvents.TAKE_ACTION, self._on_action_received)
+
+    def _update_vm_util(self):
+        """Update all live VMs memory utilization.
+        The length of VMs utilization series could be difference among all VMs,
+        because index 0 represents the VM's CPU utilization at the tick it starts.
+        """
+
+        for _, vm in self._vm.items():
+            vm.util_cpu = vm.get_util(cur_tick=self._tick)
+
+    def _update_pm_util(self):
+        """Update memory utilization occupied by total VMs on each PM."""
+        for pm in self._machines:
+            total_cpu: int = 0
+            for vm_id in pm.vm_set:
+                vm = self._vm[vm_id]
+                cur_util_cores = vm.util_cpu * vm.req_cpu / 100
+                total_cpu += cur_util_cores
+            pm.util_cpu = total_cpu / pm.cap_cpu * 100
+            pm.update_util_series(self._tick)
+
+        # TODO: Energy comsumption update.
+
+    def _on_vm_required(self, evt: CascadeEvent):
+        """Callback when there is a VM requirement generated."""
+        # Get VM data from payload.
+        payload: VmRequirementPayload = evt.payload
+        vm_req: VirtualMachine = payload.vm_req
+        buffer_time: int = payload.buffer_time
+
+        # Check all valid PMs.
+        # NOTE: Should we implement this logic inside the action scope?
+        valid_pm_list = [
+            pm.id
+            for pm in self._machines
+            if (pm.cap_cpu - (pm.cap_cpu * pm.util_cpu / 100)) >= vm_req.req_cpu
+        ]
```

Comment:
we should clearly define what's the valid _pm_list:
- all pm with enough resource
- or, all pm with enough resource and not empty + one empty pm for each pm size